### DISCUSSION TITLE
dynamic columns

### DISCUSSION THREAD

**Discussion by lazyclaw-labs** [diesel-rs/diesel#3845] [Ideas]
I've been working on linq2db in c# for quite sometime, in my opinion, diesel needs to have a couple mode features to make it easy to use.

1. code generation customization: Currently, diesel-cli generated table names and column names can't be customized. It has its own internal rules to decide what name to use. Sometimes, the column name in database does not make too much sense, either because it's an old database, or because the db designer trying to use the same column for multiple purpose to be used for multiple applications. And for your small project,  that column has a specific meaning. Linq2DB provide a template for you to customize the code generation process, give you a lot flexibility.
2. select into csv or json directly, without providing the type list when loading. due to the nature of Rust, we have to do load::<(T1, T2)>() when we want to select a subset of columns. And also, we can't serialize into json directly. We have to create a struct, marked with serde::serialize. I know why we are doing this, but can't we use  some defaults? I mean, when I do query.select(id, name, title).load(), I've already know the type/name of each column. every SQL type should have a default rust type map to (for some weird type, all map to String or Vec[u8]). We surely can dynamically construct a struct/marked as serde::serializable, thus we can return the json string to the front end directly.

I know the "dynamic" means quite different, we need to generate code ahead of compilation. But I feel this is feasible, without changing the compiler.

All in all, need more flexibility. 

What do you guys think?

**Comment by weiznich** [comment]
> code generation customization: Currently, diesel-cli generated table names and column names can't be customized. It has its own internal rules to decide what name to use. Sometimes, the column name in database does not make too much sense, either because it's an old database, or because the db designer trying to use the same column for multiple purpose to be used for multiple applications. And for your small project, that column has a specific meaning. Linq2DB provide a template for you to customize the code generation process, give you a lot flexibility.


We are happy to consider more concrete proposals for such a customization scheme. Otherwise there is currently no planned work in that direction.

> select into csv or json directly, without providing the type list when loading. due to the nature of Rust, we have to do load::<(T1, T2)>() when we want to select a subset of columns. And also, we can't serialize into json directly. We have to create a struct, marked with serde::serialize. I know why we are doing this, but can't we use some defaults? I mean, when I do query.select(id, name, title).load(), I've already know the type/name of each column. every SQL type should have a default rust type map to (for some weird type, all map to String or Vec[u8]). We surely can dynamically construct a struct/marked as serde::serializable, thus we can return the json string to the front end directly.


Checkout diesel-dynamic-schema, especially the [`DynamicRow`] type there. Its designed for that usescase, but allows you to easily adjust the mapping for your own use case by requiring that you provide the value type.

**Reply by lazyclaw-labs** [reply]
cool. I may create a PR later. :-)

**Reply by weiznich** [reply]
To prevent frustration on both sides I would prefer reading a summary of your proposed design first. That doesn't need to cover everything in detail, but you might want to explain how exactly you would like to configure such a customization feature exactly.
